design/parallelism.md

```diff
@@ -0,0 +1,41 @@
+# Dissolve Parallel Tech Spec
+
+## Introduction
+
+### Purpose
+This document lays out the current and future technical requirements for the parallel architecture of the Dissolve project.
+
+### Background
+
+Dissolve is an atomistic simulation code for the interrogation and analysis of neutron scattering data. Since the simulation boxes (collections of atoms) that it produces must be representative of the physical systems measured experimentally, particle numbers of 1 million and beyond must be treatable. This is not practical on a single-core machine, or even a modestly powerful workstation, and so suitable parallel technology must be implemented.
+
+### Current Implementation
+
+At present (v0.7.0) Dissolve utilises the widely-available OpenMPI standard to achieve parallelism across multiple CPU cores over multiple physical processing enclosures. Each slave process maintains its own copy of all simulation data (replicated data strategy) leading to high memory requirements.
+
+Dissolve's original design was to allow partitioning of processors into sub-groups in order to permit:
+
+1. Fine-grained control over partitioning of workload in order to achieve balance between processing time and communication overhead.
+2. Forking of the main loop to allow simultaneous calculations targetting different atomic configurations or modules
+
+1) is particularly important in the implementation of parallel Monte Carlo techniques, where trial moves must be distributed throughout the simulation box in such a way that they are independent of each other. This is achievable, but heavily system dependent (size, molecule type, etc.) and requires adaptive partitioning of the available processes. As a consequence of 2), each main loop iteration requires a broadcast of all simulation data between processes to ensure that each process has up-to-date information, forcing the requirement for T::broadcast() and T::equality() (enforced by inheriting from GenericItemBase) for each class object that needs to be distributed.
+
+## Requirements
+
+### Code
+
+- A single codebase must be maintained for serial, GUI and parallel versions, with parallelism built-in when requested through the build generator (CMake).
+- The GUI version will remain as a serial code, as its scope covers only the set-up and analysis of simulations, and the running of small tutorial systems that do not require to be run in parallel.
+
+### General Approach
+
+- The ability to partition available processes into task-groups for certain algorithms is beneficial.
+- Scaling over CPU- and GPU-based multi-core systems should be considered.
+- Replicated data strategies over CPU cores should not be employed - such approaches are not feasible for systems of 1 million particles and beyond.
```

Comment:
What is the bottleneck with using replicated data for large systems?  Are we memory bound, cpu bound, or IO bound?

With just a million particles, I can't imagine that the memory load would exceed a couple of gigabytes, which should be feasible for any standard workstation.

If it's CPU bound, then I would assume that there will usually be more GPU cores to split the workload across than cluster nodes.

IF it's IO bound, is it too much communication between the nodes or is each node performing too much IO to permanent storage?

I don't have the experience with trying to use EPSR on large systems like that so I don't know where it falls apart.